File: db4oj/core/src/com/db4o/MWriteNew.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o;

import com.db4o.foundation.network.YapSocket;

final class MWriteNew extends MsgObject {
    final boolean processMessageAtServer(YapSocket sock) {
        int yapClassId = _payLoad.readInt();
        YapFile stream = (YapFile)getStream();
        unmarshall(YapConst.YAPINT_LENGTH);
        synchronized (stream.i_lock) {
            YapClass yc = stream.getYapClass(yapClassId);
            _payLoad.writeEmbedded();
            stream.prefetchedIDConsumed(_payLoad.getID());
            _payLoad.address(stream.getSlot(_payLoad.getLength()));
            yc.addFieldIndices(_payLoad, true);
            stream.writeNew(yc, _payLoad);
            getTransaction().writePointer(
                _payLoad.getID(),
                _payLoad.getAddress(),
                _payLoad.getLength());
        }
        return true;
    }
}

File: db4on/core/src/core/com/db4o/MWriteNew.cs
namespace com.db4o
{
	internal sealed class MWriteNew : com.db4o.MsgObject
	{
		internal sealed override bool ProcessMessageAtServer(com.db4o.foundation.network.YapSocket
			 sock)
		{
			int yapClassId = _payLoad.ReadInt();
			com.db4o.YapFile stream = (com.db4o.YapFile)GetStream();
			Unmarshall(com.db4o.YapConst.INT_LENGTH);
			lock (stream.i_lock)
			{
				com.db4o.YapClass yc = yapClassId == 0 ? null : stream.GetYapClass(yapClassId);
				_payLoad.WriteEmbedded();
				stream.PrefetchedIDConsumed(_payLoad.GetID());
				_payLoad.Address(stream.GetSlot(_payLoad.GetLength()));
				if (yc != null)
				{
					yc.AddFieldIndices(_payLoad, true);
				}
				stream.WriteNew(yc, _payLoad);
				GetTransaction().WritePointer(_payLoad.GetID(), _payLoad.GetAddress(), _payLoad.GetLength
					());
			}
			return true;
		}
	}
}
